### ISSUE TITLE
Playmode editor support

### ISSUE THREAD

**Issue by jmickle66666666** [sabresaurus/SabreCSG#39]
More of a question about the viability rather than a feature request. Is it possible to port the editing tools to work at runtime? Probably with some prefab to place in a scene that handles the editing controls.

I'm happy to try and build this myself, but was wondering if there were any immediate gotchas or issues that would make it too impractical.

**Comment by Henry00IS** [comment]
The UnityEditor namespace can not be used in final builds of your game. All of the editing tools and gizmos are based on Unity Editor and won't work in the game. Once you google for "unity gizmos in game" things will start going downhill from there.

However, theoretically if you were to build a bunch of classes/methods that replace the missing stuff in UnityEditor you may be able to use the existing tools, but it's a stretch and probably way too complicated to mimic identical behavior (and these tools allow you so much freedom your players can build corrupt geometry).

If anything you should just consider what functionality you really need to have and possibly just copy&paste the logic to your in-game editor. Luckily the sabresaurus has left us a trail of geometry functions all over the place that will allow you to do anything.

I personally believe there is still a lot of private/internal stuff that should be made public so if you find any of those useful in real-time CSG please mark them as public and make a pull request sometime. :)

**Comment by jmickle66666666** [comment]
Thanks for all the info!

I'm aware that the gizmos and anything relying on UnityEditor would not be available but I was curious about how coupled everything is into that, as I haven't taken the time to look through myself yet. If it's "just" a case of building an interface that mimics the actions a user can take in the editor, and hooking those up to controls present in SabreCSG then that's a perfect situation but I expect there's a lot of work to decouple and find the parts that I need.

But yeah, I was expecting it would be more sensible to approach it from the other direction as you suggested: pull out the parts I want instead of building a new system on top of whats already there.

**Comment by sabresaurus** [comment]
This is something that has been requested a lot, player facing CSG tools are realistically not viable with the current state of SabreCSG. For simple CSG operations it's fairly robust but it's still pretty easy to cause lengthy builds, heavy splitting or missing/extra faces. There are some fundamental issues with how the core build code has been implemented which cause this, such as its reliance on floating point maths, but likely the brush representations too. 

The current CSG implementation (which I've nicknamed Neo) is actually the fourth that SabreCSG has had. It has significant advantages over its predecessors such as caching of parts of build data, considerably faster build times, better triangulation, support for parallelisation (although not turned on) and numerous architecture improvements based on lessons learned from the first three implementations that make it build much more rapidly.

Before I open sourced SabreCSG I was considering rewriting the core build code again with the idea of removing floating point operations completely and possibly changing some of the input representations to planes and indexed edges. Unfortunately SabreCSG never came close to supporting itself in terms of revenue and I ran out of money.

In summary, making the tools work in built games is the easy part. Making the build code robust enough that you could hand it to players is a very difficult problem.
